subsys/usb/usb_device.c

```diff
@@ -233,14 +233,28 @@ static bool usb_handle_request(struct usb_setup_packet *setup,
  *
  * @return N/A
  */
-static void usb_data_to_host(void)
+static void usb_data_to_host(u16_t len)
 {
 	u32_t chunk = usb_dev.data_buf_residue;
 
 	/*Always EP0 for control*/
-	usb_dc_ep_write(0x80, usb_dev.data_buf, chunk, &chunk);
+	usb_dc_ep_write(USB_CONTROL_IN_EP0, usb_dev.data_buf, chunk, &chunk);
 	usb_dev.data_buf += chunk;
 	usb_dev.data_buf_residue -= chunk;
+
+	/*
+	 * Windows expects a ZLP after the last IN transaction that was
+	 * EP-sized if wLength was bigger than EP size.
+	 */
+	if (!usb_dev.data_buf_residue && chunk == USB_MAX_CTRL_MPS
+	    && len > chunk) {
```

Comment:
The comment above and the commit message are confusing and not correct. It seems there is a bug in our device stack. Please change the commit message and comment to something like the one from @finikorg "Send ZLP when host asks for the bigger length and the last chunk is wMaxPacketSize to indicate about the last packet.

See also:
_The Data stage of a control transfer from an endpoint to the host is complete when the endpoint does one of
the following:
• Has transferred exactly the amount of data specified during the Setup stage
• Transfers a packet with a payload size less than wMaxPacketSize or transfers a zero-length packet
(5.5.3 USB Specification Revision 2.0)_

_A control pipe may have a variable-length data phase in which the host requests more data than is contained
in the specified data structure ....... If the data structure is an exact multiple of wMaxPacketSize for the pipe, the function will return a zero-length packet to indicate the end of the Data stage. (8.5.3.2 USB Specification Revision 2.0)_

Reply:
Agree here with @jfischer-phytec-iot

Reply:
Fixed